File: channel/src/identity/signing.ts
import * as ed from "@noble/ed25519";
import { sha512 } from "@noble/hashes/sha2.js";
import { extractPublicKey } from "./did.js";

// @noble/ed25519 v2 requires setting the hash function
ed.etc.sha512Sync = (...m) => sha512(ed.etc.concatBytes(...m));

export interface MessageEnvelope {
  from: string;
  from_did: string;
  to: string;
  to_did: string;
  type: string;
  subject: string;
  body: string;
  timestamp: string;
  from_stable_id?: string;
  to_stable_id?: string;
  message_id?: string;
  signature?: string;
  signing_key_id?: string;
}

export type VerificationStatus =
  | "verified"
  | "verified_custodial"
  | "unverified"
  | "failed"
  | "identity_mismatch";

/**
 * Build the canonical JSON payload for message signing.
 * Fields are sorted lexicographically, no whitespace, minimal escaping.
 * Optional fields (from_stable_id, message_id, to_stable_id) are omitted when empty.
 */
export function canonicalJSON(env: MessageEnvelope): string {
  const fields: [string, string][] = [
    ["body", env.body],
    ["from", env.from],
    ["from_did", env.from_did],
    ["subject", env.subject],
    ["timestamp", env.timestamp],
    ["to", env.to],
    ["to_did", env.to_did],
    ["type", env.type],
  ];

  if (env.from_stable_id) fields.push(["from_stable_id", env.from_stable_id]);
  if (env.message_id) fields.push(["message_id", env.message_id]);
  if (env.to_stable_id) fields.push(["to_stable_id", env.to_stable_id]);

  fields.sort((a, b) => (a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : 0));

  let result = "{";
  for (let i = 0; i < fields.length; i++) {
    if (i > 0) result += ",";
    result += '"' + fields[i][0] + '":"' + escapeJSON(fields[i][1]) + '"';
  }
  result += "}";
  return result;
}

/** JSON-escape a string value, matching Go's writeEscapedString exactly. */
function escapeJSON(s: string): string {
  let result = "";
  for (const ch of s) {
    const code = ch.codePointAt(0)!;
    switch (ch) {
      case '"':
        result += '\\"';
        break;
      case "\\":
        result += "\\\\";
        break;
      case "\n":
        result += "\\n";
        break;
      case "\r":
        result += "\\r";
        break;
      case "\t":
        result += "\\t";
        break;
      case "\b":
        result += "\\b";
        break;
      case "\f":
        result += "\\f";
        break;
      default:
        if (code < 0x20) {
          result += "\\u" + code.toString(16).padStart(4, "0");
        } else {
          result += ch;
        }
    }
  }
  return result;
}

function b64Encode(bytes: Uint8Array): string {
  const bin = String.fromCharCode(...bytes);
  // Base64 RFC 4648 no padding (RawStdEncoding)
  return btoa(bin).replace(/=+$/, "");
}

function b64Decode(s: string): Uint8Array {
  const bin = atob(s);
  const bytes = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) bytes[i] = bin.charCodeAt(i);
  return bytes;
}

/** Sign a message envelope. Returns base64 signature (no padding). */
export async function signMessage(
  seed: Uint8Array,
  env: MessageEnvelope,
): Promise<string> {
  const payload = canonicalJSON(env);
  const sig = ed.sign(new TextEncoder().encode(payload), seed);
  return b64Encode(sig);
}

/**
 * Verify a message envelope signature.
 * Returns 'unverified' if DID or signature is missing.
 * Returns 'failed' if signature doesn't verify.
 * Returns 'verified' if valid.
 */
export async function verifyMessage(
  env: MessageEnvelope,
): Promise<VerificationStatus> {
  if (!env.from_did || !env.signature) {
    return "unverified";
  }

  if (env.signing_key_id && env.signing_key_id !== env.from_did) {
    return "failed";
  }

  if (!env.from_did.startsWith("did:key:z")) {
    return "unverified";
  }

  let publicKey: Uint8Array;
  try {
    publicKey = extractPublicKey(env.from_did);
  } catch {
    return "failed";
  }

  let sigBytes: Uint8Array;
  try {
    sigBytes = b64Decode(env.signature);
  } catch {
    return "failed";
  }

  const payload = canonicalJSON(env);
  const valid = ed.verify(
    sigBytes,
    new TextEncoder().encode(payload),
    publicKey,
  );

  return valid ? "verified" : "failed";
}

/**
 * Verify a signature against a pre-computed canonical payload string.
 * Use when the server returns signed_payload alongside the message.
 */
export async function verifySignedPayload(
  signedPayload: string,
  signatureB64: string,
  fromDID: string,
  signingKeyID: string,
): Promise<VerificationStatus> {
  if (!fromDID || !signatureB64 || !signedPayload) {
    return "unverified";
  }

  if (signingKeyID && signingKeyID !== fromDID) {
    return "failed";
  }

  if (!fromDID.startsWith("did:key:z")) {
    return "unverified";
  }

  let publicKey: Uint8Array;
  try {
    publicKey = extractPublicKey(fromDID);
  } catch {
    return "failed";
  }

  let sigBytes: Uint8Array;
  try {
    sigBytes = b64Decode(signatureB64);
  } catch {
    return "failed";
  }

  const valid = ed.verify(
    sigBytes,
    new TextEncoder().encode(signedPayload),
    publicKey,
  );

  return valid ? "verified" : "failed";
}

File: awid/src/awid/signing.py
"""Canonical JSON payload construction plus Ed25519 signing and verification."""

from __future__ import annotations

import base64
import enum
import json

from nacl.exceptions import BadSignatureError
from nacl.signing import SigningKey, VerifyKey

from awid.did import public_key_from_did

SIGNED_FIELDS = frozenset(
    {
        "body",
        "from",
        "from_did",
        "from_stable_id",
        "message_id",
        "subject",
        "timestamp",
        "to",
        "to_did",
        "to_stable_id",
        "type",
    }
)


class VerifyResult(enum.Enum):
    VERIFIED = "verified"
    VERIFIED_CUSTODIAL = "verified_custodial"
    UNVERIFIED = "unverified"
    FAILED = "failed"


def canonical_json_bytes(fields: dict) -> bytes:
    return json.dumps(fields, sort_keys=True, separators=(",", ":"), ensure_ascii=False).encode(
        "utf-8"
    )


def canonical_payload(fields: dict) -> bytes:
    filtered = {k: v for k, v in fields.items() if k in SIGNED_FIELDS}
    return canonical_json_bytes(filtered)


def sign_message(private_key: bytes, payload: bytes) -> str:
    signing_key = SigningKey(private_key)
    signed = signing_key.sign(payload)
    return base64.b64encode(signed.signature).rstrip(b"=").decode("ascii")


def _decode_signature(signature_b64: str) -> bytes:
    padded = signature_b64 + "=" * (-len(signature_b64) % 4)
    return base64.urlsafe_b64decode(padded)


def verify_signature_with_public_key(
    public_key: bytes, payload: bytes, signature_b64: str | None
) -> VerifyResult:
    if not signature_b64:
        return VerifyResult.UNVERIFIED

    try:
        sig_bytes = _decode_signature(signature_b64)
    except Exception:
        return VerifyResult.FAILED

    try:
        verify_key = VerifyKey(public_key)
        verify_key.verify(payload, sig_bytes)
        return VerifyResult.VERIFIED
    except BadSignatureError:
        return VerifyResult.FAILED
    except Exception:
        return VerifyResult.FAILED


def verify_signature(did: str | None, payload: bytes, signature_b64: str | None) -> VerifyResult:
    if not did or not signature_b64:
        return VerifyResult.UNVERIFIED

    if not did.startswith("did:key:z"):
        return VerifyResult.UNVERIFIED

    try:
        public_key = public_key_from_did(did)
    except Exception:
        return VerifyResult.FAILED

    return verify_signature_with_public_key(public_key, payload, signature_b64)


def verify_did_key_signature(*, did_key: str, payload: bytes, signature_b64: str) -> None:
    result = verify_signature(did_key, payload, signature_b64)
    if result != VerifyResult.VERIFIED:
        raise ValueError("invalid signature")


